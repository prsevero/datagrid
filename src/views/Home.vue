<template>
    <div class="home">
        <Datagrid v-if="!isLoading" :headers="headers" :items="items" :item-keys="itemKeys" :save="save" />
        <Loading v-else />
    </div>
</template>

<script>
import Datagrid from '@/components/Datagrid.vue';
import Loading from '@/components/Loading.vue';

let href = window.location.origin;
if (href.substr(-1) !== '/') href += '/';
if (process.env.NODE_ENV === 'development') href = href.replace('8080', '8081');
const dataUrl = `${href}data`;

export default {
    name: 'home',
    data() {
        return {
            headers: [],
            isLoading: true,
            itemKeys: [
                1,
                {editable: true, key: 2},
                {fn: 'format', key: 3, type: 'date'},
                {
                    key: 4,
                    style: {width: '90px'},
                    tdStyle: {textAlign: 'right'},
                    type: 'float',
                },
            ],
            items: [],
        };
    },
    components: {Datagrid, Loading},
    methods: {
        save(data) {
            this.items = [...data];
            data = [...data];
            data.unshift(this.headers);

            (async () => {
                const response = await fetch(dataUrl, {
                    body: JSON.stringify(data),
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    method: 'POST',
                }).catch(error => {
                    // Handle error
                    // console.error(result);
                    return error;
                });
                const result = await response;

                if (result.status !== 200) {
                    // Handle error
                    // console.error(result);
                }
            })();
        },
    },
    mounted() {
        fetch(dataUrl)
            .then(response => response.json())
            .then(data => {
                this.isLoading = false;
                this.headers = data.splice(0, 1)[0];
                this.items = data;
            })
            .catch(() => {
                this.isLoading = false;
                // Handle error
                // console.error(error);
            });
    },
};
</script>
