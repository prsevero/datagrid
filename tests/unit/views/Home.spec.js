import { shallowMount } from '@vue/test-utils';
import Home from '@/views/Home.vue';
import fetch from 'node-fetch';
global.fetch = fetch;


describe('Home.vue', () => {

    let mountedHome;
    const home = () => {
        if (!mountedHome) {
            mountedHome = shallowMount(Home, {
                mocks: {
                    fetch: (url) => {
                        return Promise.resolve({data: 'testing'});
                    },
                },
            });
        }
        return mountedHome;
    }

    beforeEach(() => {
        mountedHome = undefined;
    });

    it('should match snapshot', () => {
        const wrapper = home();
        expect(wrapper.vm.$el).toMatchSnapshot();
    });

    it('always render home div', () => {
        const wrapper = home();
        expect(wrapper.element.className).toMatch('home');
        expect(wrapper.is('div')).toBe(true);
    });

    it('should fetch data when mount', done => {
        const wrapper = home();
        expect(wrapper.vm.isLoading).toBeTruthy();
        setTimeout(() => {
            expect(wrapper.vm.isLoading).toBeFalsy();
            done();
        }, 500);
    });

    it('should set items when saving data', done => {
        const data = [0, 1, 2, 3];
        const wrapper = home();
        wrapper.vm.save(data);
        setTimeout(() => {
            expect(wrapper.vm.items).toEqual(data);
            done();
        }, 500);
    });

});
