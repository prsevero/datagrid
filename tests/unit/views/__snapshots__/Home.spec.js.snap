// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue should match snapshot 1`] = `
<div
  class="home"
>
  <loading-stub />
</div>
`;
